From 7e34dbf974c71a77ca2c5a615bed70a26acf17b6 Mon Sep 17 00:00:00 2001
From: progier389 <progier@redhat.com>
Date: Fri, 4 Apr 2025 14:44:11 +0200
Subject: [PATCH] Issue 6715 - dsconf backend replication monitor fails if
 replica id starts with 0 (#6716)

* Issue 6715 - dsconf backend replication monitor fails if replica id starts with 0
lib389 fails to retrioeve the csn if the replicaid is not exactly the normalized form of a number
Typically because 010 does not match with the RUV value.
Solution: Ensure that rid get normalized before comparing them or using them in a dict

Issue: #6715

Reviewed by: @droideck, @tbordaz (Thanks!)

(cherry picked from commit d19a4f9955aa0cdf3daf1c8a5ac0f6d83bfd9a7b)
---
 .../suites/replication/regression_m2_test.py  | 129 +++++++++++++++++-
 src/lib389/lib389/__init__.py                 |   5 +-
 src/lib389/lib389/agreement.py                |   7 +-
 src/lib389/lib389/replica.py                  |  33 ++++-
 4 files changed, 163 insertions(+), 11 deletions(-)

diff --git a/dirsrvtests/tests/suites/replication/regression_m2_test.py b/dirsrvtests/tests/suites/replication/regression_m2_test.py
index ed138013d..1a2f80522 100644
--- a/dirsrvtests/tests/suites/replication/regression_m2_test.py
+++ b/dirsrvtests/tests/suites/replication/regression_m2_test.py
@@ -12,6 +12,7 @@ import time
 import logging
 import ldif
 import ldap
+import pprint
 import pytest
 import subprocess
 import time
@@ -29,7 +30,10 @@ from lib389.idm.group import Groups, Group
 from lib389.idm.domain import Domain
 from lib389.idm.directorymanager import DirectoryManager
 from lib389.idm.services import ServiceAccounts, ServiceAccount
-from lib389.replica import Replicas, ReplicationManager, ReplicaRole, BootstrapReplicationManager
+from lib389.replica import (
+    Replicas, ReplicationManager, ReplicationMonitor, ReplicaRole,
+    BootstrapReplicationManager, NormalizedRidDict
+)
 from lib389.agreement import Agreements
 from lib389 import pid_from_file
 from lib389.dseldif import *
@@ -1144,6 +1148,129 @@ def test_bulk_import(preserve_topo_m2):
     assert len(users_s1) == len(users_s2)
 
 
+def check_monitoring_status(inst):
+    creds = { 'binddn': DN_DM, 'bindpw': PW_DM }
+    repl_monitor = ReplicationMonitor(inst)
+    report_dict = repl_monitor.generate_report(lambda h,p: creds, use_json=True)
+    log.debug(f'(Monitoring status: {pprint.pformat(report_dict)}')
+
+    agmts_status = {}
+    for inst_status in report_dict.values():
+        for replica_status in inst_status:
+            suffix = replica_status['replica_root']
+            rid = replica_status['replica_id']
+            for agmt_status in replica_status['agmts_status']:
+                rag_status = agmt_status['replication-status'][0]
+                if 'Unavailable' in rag_status:
+                    aname = agmt_status['agmt-name'][0]
+                    url = f'{agmt_status["replica"][0]}/{suffix}'
+                    assert False, f"'Unavailable' found in agreement {aname} of replica {url} : {rag_status}"
+
+    assert 'Unavailable' not in str(report_dict)
+
+
+def reinit_replica(S1, S2):
+    # Reinit replication
+    agmt = Agreements(S1).list()[0]
+    agmt.begin_reinit()
+    (done, error) = agmt.wait_reinit()
+    assert done is True
+    assert error is False
+
+    repl = ReplicationManager(DEFAULT_SUFFIX)
+    repl.wait_for_replication(S1, S2)
+    repl.wait_for_replication(S2, S1)
+
+
+def test_rid_starting_with_0(topo_m2, request):
+    """Check that replication monitoring works if replica
+       id starts with 0
+
+    :id: ed0176e6-0bf7-11f0-9846-482ae39447e5
+    :setup: 2 Supplier Instances
+    :steps:
+        1. Initialize replication to ensure that init status is set
+        2. Check that monitoring status does not contains 'Unavailable'
+        3. Change replica ids to 001 and 002
+        4. Initialize replication to ensure that init status is set
+        5. Check that monitoring status does not contains 'Unavailable'
+        6. Restore the replica ids to 1 and 2
+    :expectedresults:
+        1. Success
+        2. Success
+        3. Success
+        4. Success
+        5. Success
+        6. Success
+    """
+    S1 = topo_m2.ms["supplier1"]
+    S2 = topo_m2.ms["supplier2"]
+    replicas = [ Replicas(inst).get(DEFAULT_SUFFIX) for inst in topo_m2 ]
+
+    # Reinit replication (to ensure that init status is set)
+    reinit_replica(S1, S2)
+
+    # Get replication monitoring results
+    check_monitoring_status(S1)
+
+    # Change replica id
+    for replica,rid in zip(replicas, ['010', '020']):
+        replica.replace('nsDS5ReplicaId', rid)
+
+    # Restore replica id in finalizer
+    def fin():
+        for replica,rid in zip(replicas, ['1', '2']):
+            replica.replace('nsDS5ReplicaId', rid)
+        reinit_replica(S1, S2)
+
+    request.addfinalizer(fin)
+    # Reinit replication
+    reinit_replica(S1, S2)
+
+    # Get replication monitoring results
+    check_monitoring_status(S1)
+
+
+def test_normalized_rid_dict():
+    """Check that lib389.replica NormalizedRidDict class behaves as expected
+
+    :id: 0f88a29c-0fcd-11f0-b5df-482ae39447e5
+    :setup: None
+    :steps:
+        1. Initialize a NormalizedRidDict
+        2. Check that normalization do something
+        3. Check that key stored in NormalizedRidDict are normalized
+        4. Check that normalized and non normalized keys have the same value
+    :expectedresults:
+        1. Success
+        2. Success
+        3. Success
+        4. Success
+    """
+
+    sd = { '1': 'v1', '020': 'v2' }
+    nsd = { NormalizedRidDict.normalize_rid(key): val for key,val in sd.items() }
+    nkeys = list(nsd.keys())
+
+    # Initialize a NormalizedRidDict
+    nrd = NormalizedRidDict()
+    for key,val in sd.items():
+        nrd[key] = val
+
+    # Check that normalization do something
+    assert nkeys != list(sd.keys())
+
+    # Check that key stored in NormalizedRidDict are normalized
+    for key in nrd.keys():
+        assert key in nkeys
+
+    # Check that normalized and non normalized keys have the same value
+    for key,val in sd.items():
+        nkey = NormalizedRidDict.normalize_rid(key)
+        assert nrd[key] == val
+        assert nrd[nkey] == val
+
+
 def test_online_reinit_may_hang(topo_with_sigkill):
     """Online reinitialization may hang when the first
        entry of the DB is RUV entry instead of the suffix
diff --git a/src/lib389/lib389/__init__.py b/src/lib389/lib389/__init__.py
index ac8f5f3f2..1ac9770b0 100644
--- a/src/lib389/lib389/__init__.py
+++ b/src/lib389/lib389/__init__.py
@@ -60,6 +60,7 @@ from lib389.utils import (
     normalizeDN,
     escapeDNValue,
     ensure_bytes,
+    ensure_int,
     ensure_str,
     ensure_list_str,
     format_cmd_list,
@@ -3401,7 +3402,7 @@ class DirSrv(SimpleLDAPObject, object):
                 # Error
                 consumer.close()
                 return None
-            rid = ensure_str(replica_entries[0].getValue(REPL_ID))
+            rid = ensure_int(replica_entries[0].getValue(REPL_ID))
         except:
             # Error
             consumer.close()
@@ -3418,7 +3419,7 @@ class DirSrv(SimpleLDAPObject, object):
                 return error_msg
             elements = ensure_list_str(entry[0].getValues('nsds50ruv'))
             for ruv in elements:
-                if ('replica %s ' % rid) in ruv:
+                if ('replica %d ' % rid) in ruv:
                     ruv_parts = ruv.split()
                     if len(ruv_parts) == 5:
                         return ruv_parts[4]
diff --git a/src/lib389/lib389/agreement.py b/src/lib389/lib389/agreement.py
index 170ab9050..745dd0bd0 100644
--- a/src/lib389/lib389/agreement.py
+++ b/src/lib389/lib389/agreement.py
@@ -161,7 +161,7 @@ class Agreement(DSLdapObject):
         from lib389.replica import Replicas
         replicas = Replicas(self._instance)
         replica = replicas.get(suffix)
-        rid = replica.get_attr_val_utf8(REPL_ID)
+        rid = int(replica.get_attr_val_utf8(REPL_ID))
 
         # Open a connection to the consumer
         consumer = DirSrv(verbose=self._instance.verbose)
@@ -191,12 +191,15 @@ class Agreement(DSLdapObject):
             else:
                 elements = ensure_list_str(entry[0].getValues('nsds50ruv'))
                 for ruv in elements:
-                    if ('replica %s ' % rid) in ruv:
+                    if ('replica %d ' % rid) in ruv:
                         ruv_parts = ruv.split()
                         if len(ruv_parts) == 5:
                             result_msg = ruv_parts[4]
                         break
         except ldap.INVALID_CREDENTIALS as e:
+            self._log.debug('Failed to search for the suffix ' +
+                                     '({}) consumer ({}:{}) failed, error: {}'.format(
+                                         suffix, host, port, e))
             raise(e)
         except ldap.LDAPError as e:
             self._log.debug('Failed to search for the suffix ' +
diff --git a/src/lib389/lib389/replica.py b/src/lib389/lib389/replica.py
index 07f75b878..0486c48f8 100644
--- a/src/lib389/lib389/replica.py
+++ b/src/lib389/lib389/replica.py
@@ -822,6 +822,26 @@ class ReplicaLegacy(object):
             raise ValueError('Failed to update replica: ' + str(e))
 
 
+class NormalizedRidDict(dict):
+    """A dict whose key is a Normalized Replica ID
+    """
+
+    @staticmethod
+    def normalize_rid(rid):
+        return int(rid)
+
+    def __init__(self):
+        super().__init__()
+
+    def __getitem__(self, key):
+        nkey = NormalizedRidDict.normalize_rid(key)
+        return super().__getitem__(nkey)
+
+    def __setitem__(self, key, value):
+        nkey = NormalizedRidDict.normalize_rid(key)
+        super().__setitem__(nkey, value)
+
+
 class RUV(object):
     """Represents the server in memory RUV object. The RUV contains each
     update vector the server knows of, along with knowledge of CSN state of the
@@ -839,11 +859,11 @@ class RUV(object):
         else:
             self._log = logging.getLogger(__name__)
         self._rids = []
-        self._rid_url = {}
-        self._rid_rawruv = {}
-        self._rid_csn = {}
-        self._rid_maxcsn = {}
-        self._rid_modts = {}
+        self._rid_url = NormalizedRidDict()
+        self._rid_rawruv = NormalizedRidDict()
+        self._rid_csn = NormalizedRidDict()
+        self._rid_maxcsn = NormalizedRidDict()
+        self._rid_modts = NormalizedRidDict()
         self._data_generation = None
         self._data_generation_csn = None
         # Process the array of data
@@ -935,9 +955,10 @@ class RUV(object):
         :returns: str
         """
         self._log.debug("Allocated rids: %s" % self._rids)
+        rids = [ int(rid) for rid in self._rids ]
         for i in range(1, 65534):
             self._log.debug("Testing ... %s" % i)
-            if str(i) not in self._rids:
+            if i not in rids:
                 return str(i)
         raise Exception("Unable to alloc rid!")
 
-- 
2.49.0

